tests/unit/test_request_id_header.py

```diff
@@ -0,0 +1,387 @@
+    MockServerTestBase,
+    add_select1_result,
+)
+from google.cloud.spanner_v1.testing.interceptors import XGoogRequestIDHeaderInterceptor
+from google.cloud.spanner_v1 import (
+    BatchCreateSessionsRequest,
+    BeginTransactionRequest,
+    ExecuteSqlRequest,
+)
+from google.api_core.exceptions import Aborted
+from google.rpc import code_pb2
+from google.cloud.spanner_v1.request_id_header import REQ_RAND_PROCESS_ID
+
+
+class TestRequestIDHeader(MockServerTestBase):
+    def tearDown(self):
+        self.database._x_goog_request_id_interceptor.reset()
+
+    def test_snapshot_read(self):
+        add_select1_result()
+        if not getattr(self.database, "_interceptors", None):
+            self.database._interceptors = MockServerTestBase._interceptors
+        with self.database.snapshot() as snapshot:
+            results = snapshot.execute_sql("select 1")
+            result_list = []
+            for row in results:
+                result_list.append(row)
+                self.assertEqual(1, row[0])
+            self.assertEqual(1, len(result_list))
+
+        requests = self.spanner_service.requests
+        self.assertEqual(2, len(requests), msg=requests)
+        self.assertTrue(isinstance(requests[0], BatchCreateSessionsRequest))
+        self.assertTrue(isinstance(requests[1], ExecuteSqlRequest))
+
+        # Now ensure monotonicity of the received request-id segments.
+        got_stream_segments, got_unary_segments = self.canonicalize_request_id_headers()
+        want_unary_segments = [
+            (
+                "/google.spanner.v1.Spanner/BatchCreateSessions",
+                (1, REQ_RAND_PROCESS_ID, 1, 1, 1, 1),
+            )
+        ]
+        want_stream_segments = [
+            (
+                "/google.spanner.v1.Spanner/ExecuteStreamingSql",
+                (1, REQ_RAND_PROCESS_ID, 1, 1, 2, 1),
+            )
+        ]
+
+        assert got_unary_segments == want_unary_segments
+        assert got_stream_segments == want_stream_segments
+
+    def test_snapshot_read_concurrent(self):
+        def select1():
+            with self.database.snapshot() as snapshot:
+                rows = snapshot.execute_sql("select 1")
+                res_list = []
+                for row in rows:
+                    self.assertEqual(1, row[0])
+                    res_list.append(row)
+                self.assertEqual(1, len(res_list))
+
+        n = 10
+        threads = []
+        for i in range(n):
+            th = threading.Thread(target=select1, name=f"snapshot-select1-{i}")
+            th.run()
+            threads.append(th)
+
+        random.shuffle(threads)
+
+        while True:
+            n_finished = 0
+            for thread in threads:
+                if thread.is_alive():
+                    thread.join()
+                else:
+                    n_finished += 1
+
+            if n_finished == len(threads):
+                break
+
+            time.sleep(1)
```

Comment:
nit: can we remove this? The call to `thread.join()` should ensure that the threads have finished.